### PR TITLE
pkg/download: close files after extracting in tar

### LINKED ISSUE #308: Installing some plugins broken on windows
(had to use  https://storage.googleapis.com/kubernetes-release/release/v1.17.0-alpha.0/bin/windows/amd64/kubectl.exe to be able to execute plugins)

It looks like plugins can't be installed on windows. It seems like while moving the files, there's an open file handle.

```
c:\> kubectl krew install get-all
Updated the local copy of plugin index.
Installing plugin: get-all
W0819 16:19:15.140637    5160 install.go:132] failed to install plugin "get-all": failed to dowload and move during installation: failed to move files: failed moving files: could not rename file from "C:\\Users\\alp\\AppData\\Local\\Temp\\krew-downloads\\get-all\\ketall-windows-amd64" to "C:\\Users\\alp\\AppData\\Local\\Temp\\krew-temp-move160330531\\ketall-windows-amd64.exe": rename C:\Users\alp\AppData\Local\Temp\krew-downloads\get-all\ketall-windows-amd64 C:\Users\alp\AppData\Local\Temp\krew-temp-move160330531\ketall-windows-amd64.exe: The process cannot access the file because it is being used by another process.
F0819 16:19:15.141640    5160 root.go:52] failed to install some plugins: [get-all]
exit status 255
```

<details>
<summary>full output</summary>
<pre>
I0819 16:35:19.490880    6512 scanner.go:74] Reading plugin "get-all"
I0819 16:35:19.491899    6512 install.go:119] Will install plugin: get-all
Installing plugin: get-all
I0819 16:35:19.491899    6512 install.go:74] Looking for installed versions
I0819 16:35:19.491899    6512 util.go:79] Searching for installed versions of get-all in "C:\\Users\\alp\\.krew\\bin"
I0819 16:35:19.492898    6512 install.go:83] Finding download target for plugin get-all
I0819 16:35:19.492898    6512 util.go:38] Using os=windows arch=amd64
I0819 16:35:19.492898    6512 util.go:61] Matching platform for labels(arch=amd64,os=windows)
I0819 16:35:19.492898    6512 util.go:68] Found matching platform with index (2)
I0819 16:35:19.492898    6512 util.go:131] Matching plugin version is e0313c356d280ad12a858b8de539aa1e5cb19908589483bd153896ace08e679f
I0819 16:35:19.492898    6512 install.go:46] Creating download dir "C:\\Users\\alp\\AppData\\Local\\Temp\\krew-downloads\\get-all"
I0819 16:35:19.492898    6512 install.go:61] Getting sha256 (e0313c356d280ad12a858b8de539aa1e5cb19908589483bd153896ace08e679f) signed version
I0819 16:35:19.492898    6512 downloader.go:36] Fetching "https://github.com/corneliusweig/ketall/releases/download/v1.2.0/bundle.tar.gz"
I0819 16:35:20.227011    6512 downloader.go:43] Reading download data into memory
I0819 16:35:28.668521    6512 downloader.go:48] Read 32492427 bytes of download data into memory
I0819 16:35:28.668521    6512 downloader.go:177] detected .tar.gz file
I0819 16:35:28.668521    6512 downloader.go:93] tar: extracting to "C:\\Users\\alp\\AppData\\Local\\Temp\\krew-downloads\\get-all"
I0819 16:35:28.669537    6512 downloader.go:110] tar: processing "ketall-linux-amd64" (type=48, mode=-rwxr-xr-x)
I0819 16:35:28.669537    6512 downloader.go:125] tar: ensuring parent dirs exist for regular file, dir=C:\Users\alp\AppData\Local\Temp\krew-downloads\get-all
I0819 16:35:28.954431    6512 downloader.go:139] tar: processed "ketall-linux-amd64"
I0819 16:35:28.954828    6512 downloader.go:110] tar: processing "ketall-windows-amd64" (type=48, mode=-rwxr-xr-x)
I0819 16:35:28.955833    6512 downloader.go:125] tar: ensuring parent dirs exist for regular file, dir=C:\Users\alp\AppData\Local\Temp\krew-downloads\get-all
I0819 16:35:29.241547    6512 downloader.go:139] tar: processed "ketall-windows-amd64"
I0819 16:35:29.241547    6512 downloader.go:110] tar: processing "ketall-darwin-amd64" (type=48, mode=-rwxr-xr-x)
I0819 16:35:29.243532    6512 downloader.go:125] tar: ensuring parent dirs exist for regular file, dir=C:\Users\alp\AppData\Local\Temp\krew-downloads\get-all
I0819 16:35:29.545013    6512 downloader.go:139] tar: processed "ketall-darwin-amd64"
I0819 16:35:29.545013    6512 downloader.go:141] tar extraction to C:\Users\alp\AppData\Local\Temp\krew-downloads\get-all complete
I0819 16:35:29.546163    6512 move.go:155] Creating plugin dir "C:\\Users\\alp\\.krew\\store\\get-all"
I0819 16:35:29.548205    6512 move.go:161] Creating temp plugin move operations dir "C:\\Users\\alp\\AppData\\Local\\Temp\\krew-temp-move050602455"
I0819 16:35:29.548205    6512 move.go:125] Finding move targets from "C:\\Users\\alp\\AppData\\Local\\Temp\\krew-downloads\\get-all" to "C:\\Users\\alp\\AppData\\Local\\Temp\\krew-temp-move050602455" with file operation=index.FileOperation{From:"./ketall-windows-amd64", To:"ketall-windows-amd64.exe"}
I0819 16:35:29.549210    6512 move.go:44] Trying to move single file directly from="C:\\Users\\alp\\AppData\\Local\\Temp\\krew-downloads\\get-all" to="C:\\Users\\alp\\AppData\\Local\\Temp\\krew-temp-move050602455" with file operation=index.FileOperation{From:"./ketall-windows-amd64", To:"ketall-windows-amd64.exe"}
I0819 16:35:29.572998    6512 move.go:48] Detected single move from file operation=index.FileOperation{From:"./ketall-windows-amd64", To:"ketall-windows-amd64.exe"}
I0819 16:35:29.572998    6512 move.go:132] Move file from "C:\\Users\\alp\\AppData\\Local\\Temp\\krew-downloads\\get-all\\ketall-windows-amd64" to "C:\\Users\\alp\\AppData\\Local\\Temp\\krew-temp-move050602455\\ketall-windows-amd64.exe"
W0819 16:35:29.588991    6512 install.go:132] failed to install plugin "get-all": failed to dowload and move during installation: failed to move files: failed moving files: could not rename file from "C:\\Users\\alp\\AppData\\Local\\Temp\\krew-downloads\\get-all\\ketall-windows-amd64" to "C:\\Users\\alp\\AppData\\Local\\Temp\\krew-temp-move050602455\\ketall-windows-amd64.exe": rename C:\Users\alp\AppData\Local\Temp\krew-downloads\get-all\ketall-windows-amd64 C:\Users\alp\AppData\Local\Temp\krew-temp-move050602455\ketall-windows-amd64.exe: The process cannot access the file because it is being used by another process.
F0819 16:35:29.589993    6512 root.go:50] failed to install some plugins: [get-all]
github.com/GoogleContainerTools/krew/cmd/krew/cmd.init.1.func1
        /go/src/github.com/GoogleContainerTools/krew/cmd/krew/cmd/install.go:142
github.com/GoogleContainerTools/krew/vendor/github.com/spf13/cobra.(*Command).execute
        /go/src/github.com/GoogleContainerTools/krew/vendor/github.com/spf13/cobra/command.go:762
github.com/GoogleContainerTools/krew/vendor/github.com/spf13/cobra.(*Command).ExecuteC
        /go/src/github.com/GoogleContainerTools/krew/vendor/github.com/spf13/cobra/command.go:852
github.com/GoogleContainerTools/krew/vendor/github.com/spf13/cobra.(*Command).Execute
        /go/src/github.com/GoogleContainerTools/krew/vendor/github.com/spf13/cobra/command.go:800
github.com/GoogleContainerTools/krew/cmd/krew/cmd.Execute
        /go/src/github.com/GoogleContainerTools/krew/cmd/krew/cmd/root.go:48
main.main
        /go/src/github.com/GoogleContainerTools/krew/cmd/krew/main.go:23
runtime.main
        /usr/local/go/src/runtime/proc.go:201
runtime.goexit
        /usr/local/go/src/runtime/asm_amd64.s:1333
exit status 255
</pre>
</details>

however this is not always reproducible, on another plugin, it fails because I am running on a cmd.exe without Administrator privileges:

```
c:\>kubectl krew install who-can
Updated the local copy of plugin index.
Installing plugin: who-can
W0819 16:23:05.164281    5724 install.go:132] failed to install plugin "who-can": failed to create a symlink form "C:\\Users\\alp\\.krew\\bin" to "C:\\Users\\alp\\.krew\\bin\\kubectl-who_can.exe": symlink C:\Users\alp\.krew\store\who-can\29d8c8c85db54a8a748d7ca1f3d523c5dac123cadeb80907b7f66a9cf55040d6\kubectl-who-can.exe C:\Users\alp\.krew\bin\kubectl-who_can.exe: A required privilege is not held by the client.
F0819 16:23:05.165294    5724 root.go:52] failed to install some plugins: [who-can]
exit status 255
```

However if I re-run cmd.exe as Administrator, this `who-can` installs, yet `get-all` fails with the same error as before indicating there's an open descriptor/handle on the file.

version info:
GitTag          v0.2.1
GitCommit       cd2c6e1

/area platform/windows
/priority P1
/kind bug

### PR DESCRIPTION
**Pull request by ahmetb**
- Close file handle after extracting files on windows in extractTARGZ().
- Also ensure we close when io.Copy fails on extractZIP().
- Log the ignored error from the cleanup of DownloadPath.

Fixes #308.

### CODE REVIEW

File: pkg/download/downloader.go [hunk 1]
```diff
@@ -131,9 +132,12 @@ func extractTARGZ(targetDir string, at io.ReaderAt, size int64) error {
 			if err != nil {
 				return errors.Wrapf(err, "failed to create file %q", path)
 			}
+			close := func() { f.Close() }
```
Comment (corneliusweig) [inline pkg/download/downloader.go]: Would a `defer f.Close()` not be a cleaner solution?

File: pkg/download/downloader.go [hunk 2]
```diff
@@ -74,15 +74,16 @@ func extractZIP(targetDir string, read io.ReaderAt, size int64) error {
 		if err != nil {
 			return errors.Wrap(err, "can't create file in zip destination dir")
```
Comment (corneliusweig) [inline pkg/download/downloader.go]: I guess we should `src.Close()` here?

File: pkg/download/downloader.go [hunk 3]
```diff
@@ -74,15 +74,16 @@ func extractZIP(targetDir string, read io.ReaderAt, size int64) error {
...
 		}
+		close := func() {
```
Comment (corneliusweig) [inline pkg/download/downloader.go]: I don't quite see why `defer` can't be used here.

File: pkg/download/downloader.go (hunk 2 above)
Comment (ahmetb) [reply pkg/download/downloader.go]: yeah I think this is valid.

File: pkg/download/downloader.go (hunk 3 above)
Comment (ahmetb) [reply pkg/download/downloader.go]: it would defer until all of the files are extracted.

File: pkg/download/downloader.go (hunk 1 above)
Comment (ahmetb) [reply pkg/download/downloader.go]: we could end up with have thousands of files open and end up exceeding open fd limit. no need to delay.

**Pull request merged**